File: examples/signaling-server-python/signaling-server.py
#!/usr/bin/env python
#
# Python signaling server example for libdatachannel
# Copyright (c) 2020 Paul-Louis Ageneau
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.

import sys
import ssl
import json
import asyncio
import logging
import websockets


logger = logging.getLogger('websockets')
logger.setLevel(logging.INFO)
logger.addHandler(logging.StreamHandler(sys.stdout))

clients = {}


async def handle_websocket(websocket, path):
    client_id = None
    try:
        splitted = path.split('/')
        splitted.pop(0)
        client_id = splitted.pop(0)
        print('Client {} connected'.format(client_id))

        clients[client_id] = websocket
        while True:
            data = await websocket.recv()
            print('Client {} << {}'.format(client_id, data))
            message = json.loads(data)
            destination_id = message['id']
            destination_websocket = clients.get(destination_id)
            if destination_websocket:
                message['id'] = client_id
                data = json.dumps(message)
                print('Client {} >> {}'.format(destination_id, data))
                await destination_websocket.send(data)
            else:
                print('Client {} not found'.format(destination_id))

    except Exception as e:
        print(e)

    finally:
        if client_id:
            del clients[client_id]
            print('Client {} disconnected'.format(client_id))


async def main():
    # Usage: ./server.py [[host:]port] [SSL certificate file]
    endpoint_or_port = sys.argv[1] if len(sys.argv) > 1 else "8000"
    ssl_cert = sys.argv[2] if len(sys.argv) > 2 else None

    endpoint = endpoint_or_port if ':' in endpoint_or_port else "127.0.0.1:" + endpoint_or_port

    if ssl_cert:
        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
        ssl_context.load_cert_chain(ssl_cert)
    else:
        ssl_context = None

    print('Listening on {}'.format(endpoint))
    host, port = endpoint.rsplit(':', 1)

    server = await websockets.serve(handle_websocket, host, int(port), ssl=ssl_context)
    await server.wait_closed()


if __name__ == '__main__':
    asyncio.run(main())

File: examples/signaling-server-qt/SignalingServer.cpp
/**
 * Qt signaling server example for libdatachannel
 * Copyright (c) 2022 cheungxiongwei
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

#include "SignalingServer.h"
#include <QDebug>
#include <QtWebSockets>
#include <format>

SignalingServer::SignalingServer(QObject *parent) : QObject(parent) {
	server = new QWebSocketServer("SignalingServer", QWebSocketServer::NonSecureMode, this);
	QObject::connect(server, &QWebSocketServer::newConnection, this,
	                 &SignalingServer::onNewConnection);
}

bool SignalingServer::listen(const QHostAddress &address, quint16 port) {
	return server->listen(address, port);
}

void SignalingServer::onNewConnection() {
	auto webSocket = server->nextPendingConnection();
	auto client_id = webSocket->requestUrl().path().split("/").at(1);
	qInfo() << QString::fromStdString(
	    std::format("Client {} connected", client_id.toUtf8().constData()));

	clients[client_id] = webSocket;

	webSocket->setObjectName(client_id);
	QObject::connect(webSocket, &QWebSocket::disconnected, this, &SignalingServer::onDisconnected);
	QObject::connect(webSocket, QOverload<QAbstractSocket::SocketError>::of(&QWebSocket::error),
	                 this, &SignalingServer::onWebSocketError);
	QObject::connect(webSocket, &QWebSocket::binaryMessageReceived, this,
	                 &SignalingServer::onBinaryMessageReceived);
	QObject::connect(webSocket, &QWebSocket::textMessageReceived, this,
	                 &SignalingServer::onTextMessageReceived);
}

void SignalingServer::onDisconnected() {
	QWebSocket *webSocket = qobject_cast<QWebSocket *>(sender());
	clients.remove(webSocket->objectName());
}

void SignalingServer::onWebSocketError(QAbstractSocket::SocketError error) {
	qDebug() << QString::fromStdString(std::format("Client {} << {}",
	                                               sender()->objectName().toUtf8().constData(),
	                                               QString::number(error).toUtf8().constData()));
}

void SignalingServer::onBinaryMessageReceived(const QByteArray &message) {
	qInfo() << QString::fromStdString(std::format(
	    "Client {} << {}", sender()->objectName().toUtf8().constData(), message.constData()));
}

void SignalingServer::onTextMessageReceived(const QString &message) {
	QWebSocket *webSocket = qobject_cast<QWebSocket *>(sender());

	qInfo() << QString::fromStdString(std::format("Client {} << {}",
	                                              webSocket->objectName().toUtf8().constData(),
	                                              message.toUtf8().constData()));

	auto JsonObject = QJsonDocument::fromJson(message.toUtf8()).object();
	auto destination_id = JsonObject["id"].toString();
	auto destination_websocket = clients[destination_id];
	if (destination_websocket) {
		JsonObject["id"] = webSocket->objectName();
		auto data = QJsonDocument(JsonObject).toJson(QJsonDocument::Compact);
		qInfo() << QString::fromStdString(
		    std::format("Client {} >> {}", destination_id.toUtf8().constData(), data.constData()));
		destination_websocket->sendTextMessage(QString(data));
		destination_websocket->flush();
	} else {
		qInfo() << QString::fromStdString(
		    std::format("Client {} not found", destination_id.toUtf8().constData()));
	}
}
